### PR TITLE
expression, planner: remove some fields from `expression.Column`

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

- To strengthen the meaning for `UniqueID`
- For the case that there're a lot of `UNION`, the CPU usage of `Schema` and `Column` is around 25%. This pr enables possibility to optimize it.
- Make it possible to not call `schema.Clone` and `Column.Clone` in the future.
- Clean up fields.

### What is changed and how it works?

- Clean the fields `XXXXName` in `expression.Column`. They don't affect the optimization logic.
- Remove the field `IsReferenced`. It's introduced by expression rewriter. We can remove it by simply reset the name to empty.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -131,11 +139,24 @@ func (b *PlanBuilder) getExpressionRewriter(ctx context.Context, p LogicalPlan)
...
 	rewriter.ctx = ctx
 	return
 }
 
 func (b *PlanBuilder) rewriteExprNode(rewriter *expressionRewriter, exprNode ast.ExprNode, asScalar bool) (expression.Expression, LogicalPlan, error) {
+	if rewriter.p != nil {
+		curColLen := rewriter.p.Schema().Len()
+		defer func() {
+			names := rewriter.p.OutputNames()
+			names = names[:curColLen]
```
Comment (lzmhhh123) [inline planner/core/expression_rewriter.go]: merge the two lines to `names := rewriter.p.OutputNames()[:curColLen]`

File: planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -1057,7 +1108,9 @@ func (er *expressionRewriter) binaryOpToExpression(v *ast.BinaryOperationExpr) {
 		return
 	}
 	er.ctxStack = er.ctxStack[:stkLen-2]
+	er.ctxNameStk = er.ctxNameStk[:stkLen-2]
```
Comment (lzmhhh123) [inline planner/core/expression_rewriter.go]: May we think about merging the `ctxStack` and `ctxNameStl`? They always do the same operation.

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -725,34 +758,51 @@ func (b *PlanBuilder) buildProjectionFieldNameFromExpressions(ctx context.Contex
...
 	col, isCol := expr.(*expression.Column)
 	// Correlated column won't affect the final output names. So we can put it in any of the three logic block.
 	// Don't put it into the first block just for simplifying the codes.
 	if colNameField, ok := innerNode.(*ast.ColumnNameExpr); ok && isCol {
 		// Field is a column reference.
-		colName, origColName, tblName, origTblName, dbName = b.buildProjectionFieldNameFromColumns(field, colNameField, col)
+		idx := p.Schema().ColumnIndex(col)
+		var name *types.FieldName
+		if idx == -1 {
+			ids := make([]int64, 0, p.Schema().Len())
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/logical_plan_builder.go]: Oh this is used for debug.

File: planner/core/expression_rewriter.go (hunk 2 above)
Comment (winoros) [reply planner/core/expression_rewriter.go]: I'll try this way.

Comment (winoros) [reply planner/core/expression_rewriter.go]: changed

File: go.mod [hunk 4]
```diff
@@ -40,7 +40,7 @@ require (
 	github.com/pingcap/goleveldb v0.0.0-20171020122428-b9ff6c35079e
 	github.com/pingcap/kvproto v0.0.0-20190910074005-0e61b6f435c1
 	github.com/pingcap/log v0.0.0-20190715063458-479153f07ebd
-	github.com/pingcap/parser v0.0.0-20191011021308-7586d610b7aa
```
[the remark this answers is not in the document]
Comment (winoros) [reply go.mod]: use parser's new master

File: planner/core/expression_rewriter.go [hunk 5]
```diff
@@ -301,23 +350,26 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (ast.Node, bool) {
...
 			return inNode, false
 		}
-		if col == nil {
+		if idx < 0 {
 			er.err = ErrUnknownColumn.GenWithStackByArgs(v.Column.Name.OrigColName(), "field list")
 			return inNode, false
 		}
+		col := schema.Columns[idx]
 		er.ctxStack = append(er.ctxStack, expression.NewValuesFunc(er.sctx, col.Index, col.RetType))
+		er.ctxNameStk = append(er.ctxNameStk, types.EmptyName)
```
Comment (lzmhhh123) [inline planner/core/expression_rewriter.go]: Why not use `ctxStackAppend`?

Comment (winoros) [reply planner/core/expression_rewriter.go]: Oh it should be a mistake.

File: planner/core/expression_rewriter.go [hunk 6]
```diff
@@ -329,6 +380,7 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (ast.Node, bool) {
 			return inNode, true
 		}
 		er.ctxStack = append(er.ctxStack, er.schema.Columns[index])
+		er.ctxNameStk = append(er.ctxNameStk, er.names[index])
```
Comment (lzmhhh123) [inline planner/core/expression_rewriter.go]: Ditto.

File: planner/core/planbuilder.go [hunk 7]
```diff
@@ -1019,27 +1011,33 @@ func (b *PlanBuilder) buildCheckIndexSchema(tn *ast.TableName, indexName string)
...
+			names = append(names, &types.FieldName{
+				ColName: idxCol.Name,
+				TblName: tn.Name,
+				DBName:  tn.Schema,
+			})
 			schema.Append(&expression.Column{
-				ColName:  idxCol.Name,
-				TblName:  tn.Name,
-				DBName:   tn.Schema,
+
```
Comment (lzmhhh123) [inline planner/core/planbuilder.go]: Redundant blank line.

File: planner/core/planbuilder.go [hunk 8]
```diff
@@ -1019,27 +1011,33 @@ func (b *PlanBuilder) buildCheckIndexSchema(tn *ast.TableName, indexName string)
...
+		names = append(names, &types.FieldName{
+			ColName: model.NewCIStr("extra_handle"),
+			TblName: tn.Name,
+			DBName:  tn.Schema,
+		})
 		schema.Append(&expression.Column{
-			ColName:  model.NewCIStr("extra_handle"),
-			TblName:  tn.Name,
-			DBName:   tn.Schema,
+
```
Comment (lzmhhh123) [inline planner/core/planbuilder.go]: Ditto.

**Pull request merged**